### PR TITLE
[Rogue] Updates to the support for sepsis in assassination spec

### PR DESCRIPTION
**Pull request by juemrami**
### Description
Finished implementation initially started by Whispyr.

Includes:
 - Sepsis Normalizer module
 - New assassination rogue specific Sepsis spell cooldown module 
 - Improved clarity on snapshot module for tracking garrotes improved through sepsis buffs.
 - Thistle Tea logic for catching split-second casts overcapping very small amounts of energy 
### Tests
Example of improper usage of the the buffs, and no Shiv checks since combatant is not running "Lightweight Shiv" talent.
`report/Mvn3tm9X7D1JHFhb/38-Normal+The+Vigilant+Steward,+Zskarn+-+Kill+(3:23)/Sneakboii/standard/overview`

Example of exemptions for Sepsis' cast near the end of fight (last usage)
`report/NQWfRnBLMbw9gqZz/15-Normal+The+Vigilant+Steward,+Zskarn+-+Kill+(4:45)/Bigsexsee/standard/overview`

- Screenshot(s):
sepsis-snapshot-example
spesis-cooldown-example

### CODE REVIEW

File: src/analysis/retail/rogue/assassination/CHANGELOG.tsx [hunk 1]
```diff
@@ -1,14 +1,16 @@
 import { change, date } from 'common/changelog';
-import { ToppleTheNun, Bigsxy} from 'CONTRIBUTORS';
+import { ToppleTheNun, Bigsxy, Whispyr} from 'CONTRIBUTORS';
 import SpellLink from 'interface/SpellLink';
 import SPELLS from 'common/SPELLS/rogue';
 import TALENTS from 'common/TALENTS/rogue';
 import SHARED_CHANGELOG from 'analysis/retail/rogue/shared/CHANGELOG';
 
 export default [
+  change(date(2023, 8, 24), `Improved support for Sepsis with Coooldown breakdown + Snapshoting`, Bigsxy),
```
Comment (ToppleTheNun) [inline src/analysis/retail/rogue/assassination/CHANGELOG.tsx]: You could dual-attribute this change and remove the change further down:
```suggestion
  change(date(2023, 8, 24), `Improved support for Sepsis with Coooldown breakdown + Snapshoting`, [Bigsxy, Whispyr]),
```

File: src/analysis/retail/rogue/assassination/modules/spells/GarroteUptimeAndSnapshots.tsx [hunk 2]
```diff
@@ -92,6 +112,7 @@ export default class GarroteUptimeAndSnapshots extends DotSnapshots {
     if (wasUnacceptableDowngrade) {
       snapshotPerformance = QualitativePerformance.Fail;
       snapshotSummary = <div>Unacceptable downgrade of snapshot</div>;
+      console.debug(snapshots, prevSnapshots);
```
Comment (ToppleTheNun) [inline src/analysis/retail/rogue/assassination/modules/spells/GarroteUptimeAndSnapshots.tsx]: stray `console.debug`

File: src/analysis/retail/rogue/assassination/normalizers/SepsisLinkNormalizer.ts [hunk 3]
```diff
@@ -0,0 +1,163 @@
...
+    linkingEventType: [EventType.ApplyBuff, EventType.ApplyDebuff],
+    linkingEventId: [SPELLS.SEPSIS_BUFF.id, SPELLS.SEPSIS_DEBUFF.id],
+    linkRelation: FROM_HARDCAST,
+    reverseLinkRelation: INSTANT_APPLICATION,
+    referencedEventId: TALENTS.SEPSIS_TALENT.id,
+    referencedEventType: EventType.Cast,
+    forwardBufferMs: CAST_BUFFER_MS,
+    backwardBufferMs: CAST_BUFFER_MS,
+    anyTarget: true,
+    isActive: isUsingSepsis,
```
Comment (ToppleTheNun) [inline src/analysis/retail/rogue/assassination/normalizers/SepsisLinkNormalizer.ts]: Should be able to drop these `isActive` if you set the analyzer module to be active/inactive using the same method.

File: src/analysis/retail/rogue/assassination/normalizers/SepsisLinkNormalizer.ts [hunk 4]
```diff
@@ -0,0 +1,163 @@
...
+    // Sepsis Buffs/Debuffs both last 10s
+    backwardBufferMs: SEPSIS_DEBUFF_DURATION + BUFF_DROP_BUFFER,
+    maximumLinks: 1,
+    isActive: isUsingSepsis,
+  },
+];
+
+export default class SepsisLinkNormalizer extends EventLinkNormalizer {
+  constructor(options: Options) {
+    super(options, EVENT_LINKS);
```
Comment (ToppleTheNun) [inline src/analysis/retail/rogue/assassination/normalizers/SepsisLinkNormalizer.ts:98]: If you add `this.active = this.selectedCombatant.hasTalent(TALENTS.SEPSIS_TALENT)` in the constructor, you should be able to drop the `isActive` in the links themselves.

File: src/analysis/retail/rogue/assassination/modules/talents/Sepsis.tsx [hunk 5]
```diff
@@ -0,0 +1,482 @@
...
+    );
+    this.addEventListener(Events.fightend, this.onEnd);
+  }
+
+  description(): ReactNode {
+    return (
+      <>
+        <ExplanationSection>
+          <p>
+            <Trans id="guide.rogue.assassination.sections.cooldowns.sepsis.explanation">
```
Comment (ToppleTheNun) [inline src/analysis/retail/rogue/assassination/modules/talents/Sepsis.tsx]: I suggest not using the lingui components here, as we're making an effort to reduce the i18n in the spec modules. You can just remove the `<Trans id="guide.rogue.assassination.sections.cooldowns.sepsis.explanation">` wrapper and it should be fine.

File: src/analysis/retail/rogue/assassination/modules/talents/Sepsis.tsx [hunk 6]
```diff
@@ -0,0 +1,482 @@
...
+  }
+
+  private buffUsagePerformance(
+    cast: SepsisCast,
+    buffId: keyof SepsisCast['buffs'],
+  ): UsageInfo | undefined {
+    const summaryForBuff = {
+      [PRIMARY_BUFF_KEY]: (
+        <>
+          Consume the first <SpellLink spell={SPELLS.SEPSIS_BUFF} /> buff with{' '}
```
Comment (ToppleTheNun) [inline src/analysis/retail/rogue/assassination/modules/talents/Sepsis.tsx]: This is a bit lengthy for a summary; it's a bit hard to simplify it, but maybe something like:
```tsx
<>
  First buff consumed by <SpellLink spell={SPELLS.GARROTE} />
</>
```

File: src/analysis/retail/rogue/assassination/modules/talents/Sepsis.tsx [hunk 7]
```diff
@@ -0,0 +1,482 @@
...
+          <SpellLink spell={SPELLS.GARROTE} /> as early as possible.
+        </>
+      ) as React.ReactNode,
+      [SECONDARY_BUFF_KEY]: (
+        <>
+          Consume the second <SpellLink spell={SPELLS.SEPSIS_BUFF} /> buff with{' '}
```
Comment (ToppleTheNun) [inline src/analysis/retail/rogue/assassination/modules/talents/Sepsis.tsx]: This is also a bit lengthy. In a similar vein to the previous, maybe:
```tsx
<>
  Second buff consumed by <SpellLink spell={SPELLS.GARROTE} />
</>
```

File: src/analysis/retail/rogue/assassination/modules/talents/Sepsis.tsx [hunk 8]
```diff
@@ -0,0 +1,482 @@
...
+            using the buff to cast or extend an empowered <SpellLink spell={SPELLS.GARROTE} />.
+          </>
+        );
+      } else {
+        buffPerformance = QualitativePerformance.Perfect;
+        usageDetails[buffId] = (
+          <>
+            You cast <SpellLink spell={SPELLS.GARROTE} /> to consume the {firstOrSecond}{' '}
+            <SpellLink spell={SPELLS.SEPSIS_BUFF} /> buff with{' '}
+            {formatSeconds(buff.timeRemainingOnRemoval, 2)} seconds remaining it's duration.
```
Comment (ToppleTheNun) [inline src/analysis/retail/rogue/assassination/modules/talents/Sepsis.tsx]: `remaining on its duration`

File: src/analysis/retail/rogue/assassination/modules/talents/Sepsis.tsx [hunk 9]
```diff
@@ -0,0 +1,482 @@
...
+        // If the applied garrote will outlast (ie its not "full") the fight then disregard any "early" or "late" consume rules
+        const expectedGarroteEnd = buff.consumeCast.timestamp + GARROTE_BASE_DURATION;
+        const willBeFullGarrote = expectedGarroteEnd < this.owner.fight.end_time;
+
+        // Primary buff specific checks
+        if (buffId === PRIMARY_BUFF_KEY) {
+          const isLateConsume = buff.timeRemainingOnRemoval < 5 * 1000;
+          const isOpenerCast = isInOpener(buff.consumeCast, this.owner.fight);
+          if (isLateConsume && !isOpenerCast && willBeFullGarrote) {
+            buffPerformance = QualitativePerformance.Good;
```
Comment (ToppleTheNun) [inline src/analysis/retail/rogue/assassination/modules/talents/Sepsis.tsx:253]: Is this a `Good` performance if the usage details slightly further down says `This is okay`?

File: src/analysis/retail/rogue/assassination/modules/talents/Sepsis.tsx [hunk 10]
```diff
@@ -0,0 +1,482 @@
...
+    }
+    return {
+      performance: buffPerformance,
+      summary: <div>{summaryForBuff[buffId]}</div>,
+      details: <div>{usageDetails[buffId]}</div>,
+    };
+  }
+
+  private shivPerformance(cast: SepsisCast): UsageInfo | undefined {
+    const shivSummary: React.ReactNode = (
```
Comment (ToppleTheNun) [inline src/analysis/retail/rogue/assassination/modules/talents/Sepsis.tsx:308]: This summary is also a bit long; maybe the below and shift the remainder into the details view:
```tsx
<div>
  Cast a full-length <SpellLink spell={SPELLS.SHIV} /> 
</div>
```

File: src/analysis/retail/rogue/assassination/modules/talents/Sepsis.tsx [hunk 11]
```diff
@@ -0,0 +1,482 @@
...
+      cast,
+      'InstantAuraApplication',
+    );
+    const delayedBuffApplication = getRelatedBuffApplicationFromHardcast(
+      cast,
+      'DelayedAuraApplication',
+    );
+    const debuffApplication = getDebuffApplicationFromHardcast(cast);
+
+    [initialBuffApplication, delayedBuffApplication, debuffApplication].forEach((application) => {
```
Comment (ToppleTheNun) [inline src/analysis/retail/rogue/assassination/modules/talents/Sepsis.tsx]: I'd suggest making this a bit simpler by doing:
```tsx
[initialBuffApplication, delayedBuffApplication, debuffApplication].filter(isDefined).forEach((application) =>
```

Then you can remove one layer of nesting and make this easier to read.

**Review by ToppleTheNun — CHANGES_REQUESTED** [review]
The only things that _need_ to be fixed to get this in are below, otherwise I have a few suggestions:
- remove the stray `console.debug`
- remove the lingui wrapper in Sepsis.tsx

File: src/analysis/retail/rogue/assassination/modules/spells/GarroteUptimeAndSnapshots.tsx (hunk 2 above)
Comment (juemrami) [reply src/analysis/retail/rogue/assassination/modules/spells/GarroteUptimeAndSnapshots.tsx]: oh man, embarrassing xD

File: src/analysis/retail/rogue/assassination/CHANGELOG.tsx (hunk 1 above)
Comment (juemrami) [reply src/analysis/retail/rogue/assassination/CHANGELOG.tsx]: Got it, the earlier entry was there from when i branched @Whispyrs WIP, i wasnt sure if i should have kept it in or not.

**Comment by juemrami** [comment]
> The only things that _need_ to be fixed to get this in are below, otherwise I have a few suggestions:
> 
>     * remove the stray `console.debug`
> 
>     * remove the lingui wrapper in Sepsis.tsx

addressed in https://github.com/WoWAnalyzer/WoWAnalyzer/pull/6331/commits/f45ca849b73842f44cf0f4824cf46d17b01a667a and https://github.com/WoWAnalyzer/WoWAnalyzer/pull/6331/commits/233bb8637a96c406a084040d55d7041c9df4d608 respectively.

**Review by ToppleTheNun — APPROVED** [review]

**Pull request merged**